ansible/roles/csm.rr.ceph_zoning/files/ceph_haproxy.sh

```diff
@@ -25,53 +25,73 @@
+cp /etc/ceph/ceph_conf_min /etc/ceph/ceph.conf || fail "Failed to copy ceph configuration to /etc/ceph/ceph.conf"
+
+log "Generating temporary haproxy config"
+haproxy_temp_file="/etc/haproxy/haproxy_temp.cfg"
+srv/cray/scripts/metal/generate_haproxy_cfg.sh > "$haproxy_temp_file" || fail "Failed to generate haproxy config"
+
+log "Parsing generated haproxy config"
+hosts_found=false
+
+sections=("mgr:3000:grafana-backend" "rgw0:8080:rgw-backend" "mgr:8443:dashboard_back_ssl")
+for section in "${sections[@]}"; do
+    svc=$(echo "$section" | cut -d':' -f1)
+    port=$(echo "$section" | cut -d':' -f2)
+    label=$(echo "$section" | cut -d':' -f3)
+
+    log "Processing section: $label (svc=$svc, port=$port)"
+
+    hosts=$(sed -n "s/.*server server-\([a-z0-9\-]*\)-$svc :$port check weight 100.*/\1/p" "$haproxy_temp_file") || fail "Failed to parse hosts for $label"
+
+    if [ -z "$hosts" ]; then
+        log "No entries with missing IPs found for $label"
+        continue
+    else
+        log "$label contains entries with missing IPs for hosts - $hosts"
+        hosts_found=true
+    fi
 
-    hosts=$(sed -n 's/.*server server-\([a-z0-9\-]*\)-mgr :8443 check weight 100.*/\1/p' "$haproxy_temp_file")
-    echo "dashboard_back_ssl entries in haproxy config that are missing IPs are - \"${hosts}\""
     for host in $hosts; do
-      ip=$(ceph mon dump | grep ${host} | awk '{ print $2 }' |  awk -F'[:,/]' '{print $2}')
-      old_pattern="server server-$host-mgr :8443 check weight 100"
-      new_pattern="server server-$host-mgr $ip:8443 check weight 100"
-      echo "Updated entry with IP for dashboard_back_ssl section - \"${new_pattern}\""
-      sed -i "s|$old_pattern|$new_pattern|" "$haproxy_temp_file"
-    done
+        ip=$(ceph mon dump | grep "${host}" | awk '{ print $2 }' | awk -F'[:,/]' '{print $2}') || fail "Failed to get IP for host $host"
+
+        if [ -z "$ip" ]; then
+            fail "Could not resolve IP for host $host"
+        fi
+
+        old_pattern="server server-$host-$svc :$port check weight 100"
+        new_pattern="server server-$host-$svc ${ip}:$port check weight 100"
+
+        log "Updating $label config for host $host: $old_pattern -> $new_pattern"
 
-mv $haproxy_temp_file /etc/haproxy/haproxy.cfg
-systemctl enable haproxy.service
-systemctl restart haproxy.service
-EOF
+        sed -i "s|$old_pattern|$new_pattern|" "$haproxy_temp_file" || fail "Failed to update haproxy config for $host"
+    done
 done
+
+if [ "$hosts_found" = true ]; then
+    log "Replacing haproxy.cfg with updated config"
+    mv "$haproxy_temp_file" /etc/haproxy/haproxy.cfg || fail "Failed to replace haproxy.cfg"
+
+    log "Enabling haproxy.service"
+    systemctl enable haproxy.service || fail "Failed to enable haproxy.service"
+
+    log "Restarting haproxy.service"
+    systemctl restart haproxy.service || fail "Failed to restart haproxy.service"
+
+    log "HAProxy config update completed successfully!"
+fi
```

Comment:
Can slightly simplify this by just exiting when there were no hosts found.
```suggestion
[ "$hosts_found" = true ] || exit 0

log "Replacing haproxy.cfg with updated config"
mv "$haproxy_temp_file" /etc/haproxy/haproxy.cfg || fail "Failed to replace haproxy.cfg"

log "Enabling haproxy.service"
systemctl enable haproxy.service || fail "Failed to enable haproxy.service"

log "Restarting haproxy.service"
systemctl restart haproxy.service || fail "Failed to restart haproxy.service"

log "HAProxy config update completed successfully!"
```